arch/xtensa/src/esp32/esp32_wlan.c

```diff
@@ -307,76 +361,204 @@ static inline bool esp_isfreebuffer(FAR struct esp_dev_s *priv)
-   */
+  irqstate_t flags;
+  sq_entry_t *entry;
+  struct wlan_rxbuf *rxbuf;
+
+  flags = enter_critical_section();
 
-  if (!esp_isfreebuffer(priv))
+  entry = sq_remfirst(&priv->rxb);
+  if (entry)
+    {
+      rxbuf = container_of(entry, struct wlan_rxbuf, entry);
+    }
+  else
     {
-      wlerr("ERROR: No free buffers\n");
-      return -ENOMEM;
+      rxbuf = NULL;
     }
 
-  /* Check if any errors are reported in the frame */
+  leave_critical_section(flags);
+
+  return rxbuf;
+}
 
-  if (buffer == NULL || buffer_len == 0)
+/****************************************************************************
+ * Name: wlan_tx_done
+ *
+ * Description:
+ *   TxDone callback function type.
+ *
+ * Input Parameters:
+ *   ifidx  - The interface id that the tx callback has been triggered from.
+ *   data   - Pointer to the data transmitted.
+ *   len    - Length of the data transmitted.
+ *   status - True if data was transmitted sucessfully or false if failed.
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static void wlan_tx_done(uint8_t ifidx, uint8_t *data,
+                         uint16_t *len, bool status)
+{
+  FAR struct wlan_priv_s *priv = &g_wlan_priv; 
+
+  wd_cancel(&priv->txtimeout);
+
+  wlan_txavail(&priv->dev);
+}
+
+/****************************************************************************
+ * Function: wlan_rx_done
+ *
+ * Description:
+ *   This function should be called when a packet is ready to be read
+ *   from the interface. It uses the function low_level_input() that
+ *   should handle the actual reception of bytes from the network
+ *   interface. Then the type of the received packet is determined and
+ *   the appropriate input function is called.
+ *
+ * Input Parameters:
+ *   buffer - WiFi receive buffer
+ *   len    - Length of receive buffer
+ *   eb     - WiFi receive callback input eb pointer
+ *
+ * Returned Value:
+ *   OK on success; a negated errno on failure
+ *
+ ****************************************************************************/
+
+static int wlan_rx_done(void *buffer, uint16_t len, void *eb)
+{
+  struct wlan_rxbuf *rxbuf;
+  irqstate_t flags;
+  FAR struct wlan_priv_s *priv = &g_wlan_priv;
+
+  if (len > WLAN_BUF_SIZE)
     {
-      return -EAGAIN;
+      wlerr("ERROR: Wlan receive %d larger than %d\n",
+            len, WLAN_BUF_SIZE);
     }
 
-  return OK;
+  rxbuf = wlan_alloc_buffer(priv);
+  if (!rxbuf)
+    {
+      if (eb)
+        {
+          esp_wifi_free_eb(eb);
+        }
+  
+      return -ENOBUFS;
+    }
+
+  memcpy(rxbuf->buffer, buffer, len);
+  rxbuf->len = len;
+
+  if (eb)
+    {
+      esp_wifi_free_eb(eb);
+    }
+
+  flags = enter_critical_section();
+  sq_addlast(&rxbuf->entry, &priv->rxb);
+  leave_critical_section(flags);
+
+  if (work_available(&priv->rxwork))
+    {
+      work_queue(WLAN_WORK, &priv->rxwork, wlan_rxpoll, priv, 0);
+    }
+
+  return 0;
```

Comment:
```suggestion
  return OK;
```

Reply:
This is callback function and return code is passed to internal WiFi function, so I should keep 0 here.